# R scripts for TitanCNA analysis
## Contents
* [Standard Whole Genome/Exome Sequencing Analysis](#wgs)
* [10X Genomics Whole Genome Sequencing Analysis](#tenx)

## 1. Standard Whole Genome/Exome Sequencing Analysis
R script (`titanCNA.R`) for running TitanCNA analysis on standard whole genome and exome sequencing data. 

**Input files**  
  This script assumes that the necessary input files have been generated. These are generated by the KRONOS workflow. 
  1. GC-corrected, normalized read coverage using the HMMcopy suite  
    * For exome analysis, please use the `targetedSequence` argument to specify the dataframe containing the exon baits input from a bed file.  
    
    exons <- read.delim("exon_baits.bed", header = TRUE, as.is = TRUE)  
    correctReadDepth(tumWig, normWig, gcWig, mapWig, genomeStyle = "NCBI", targetedSequence = exons)  
    
  2. Tumour allelic read counts at heterozygous SNPs (identifed from the normal sample).

**Running the R script**  
1. Look at the usage of the R script  
  ```
  # from the command line
  > Rscript titanCNA.R --help
  Usage: Rscript titanCNA.R [options]


  Options:
        --id=ID
                Sample ID

        --hetFile=HETFILE
                File containing allelic read counts at HET sites. (Required)

        --cnFile=CNFILE
                File containing normalized coverage as log2 ratios. (Required)

        --outDir=OUTDIR
                Output directory to output the results. (Required)

        --numClusters=NUMCLUSTERS
                Number of clonal clusters. (Default: 1)

        --numCores=NUMCORES
                Number of cores to use. (Default: 1)

        --ploidy_0=PLOIDY_0
                Initial ploidy value; float (Default: 2)

        --estimatePloidy=ESTIMATEPLOIDY
                Estimate ploidy; TRUE or FALSE (Default: TRUE)

        --normal_0=NORMAL_0
                Initial normal contamination (1-purity); float (Default: 0.5)

        --estimateNormal=ESTIMATENORMAL
                Estimate normal contamination method; string {'map', 'fixed'} (Default: map)

        --maxCN=MAXCN
                Maximum number of copies to model; integer (Default: 8)
        (... additional arguments)
  ```

  Additional arguments to consider are the following:  
  These arguments can be used to tune the model based on variance in the read coverage data and data-type (whole-exome sequencing or whole-genome sequencing).
  ```
  --alphaK=ALPHAK
              Hyperparameter on Gaussian variance; for WES, use 1000; for WGS, use 10000; 
              float (Default: 10000)

  --alphaKHigh=ALPHAKHIGH
              Hyperparameter on Gaussian variance for extreme copy number states; 
              for WES, use 1000; for WGS, use 10000; float (Default: 10000)
  ```
2. Example usage of R script
  ```
  # normalized coverage file: test.cn.txt
  # allelic read count file: test.het.txt
  Rscript titanCNA.R --id test --hetFile test.het.txt --cnFile test.cn.txt \
    --numClusters 1 --numCores 1 --normal_0 0.5 --ploidy_0 2 \
     --chrs "c(1:22, \"X\")" --estimatePloidy TRUE --outDir ./
  ``` 

3. Running TitanCNA for multiple restarts and model selection
  `titanCNA.R` should be run with multiple restarts for different values of (a) Ploidy (2,3,4) and (b) Number of clonal clusters. This will lead to multiple solutions. Each set of solutions for a given initialization of ploidy value will be saved to a directory (e.g. run_ploidy2, run_ploidy3, run_ploidy4).
  The R script `selectSolution.R` will help select the optimal cluster from all these solutions.  The output is a 
  ```
  numClusters=3
  numCores=4
  ## run TITAN for each ploidy (2,3,4) and clusters (1 to numClusters)
  echo "Maximum number of clusters: $numClusters";
  for ploidy in $(seq 2 4)
    do
      echo "Running TITAN for $i clusters.";
      outDir=run_ploidy$ploidy
      mkdir $outDir
      for numClust in $(seq 1 $numClusters)
      do
        echo "Running for ploidy=$ploidy";
        Rscript titanCNA_v1.10.1.R --id test --hetFile test.het.txt --cnFile test.cn.txt \
        --numClusters $numClust --numCores $numCores --normal_0 0.5 --ploidy_0 $ploidy \
        --chrs "c(1:22, \"X\")" --estimatePloidy TRUE --outDir $outDir
      done
      echo "Completed job for $numClust clusters."
    done
  
  ## select optimal solution
  Rscript selectSolution.R --ploidyRun2=run_ploidy2 --ploidyRun3=run_ploidy3 --ploidyRun4=run_ploidy4 --threshold=0.05 --outFile optimalClusters.txt
  ```

  
## 2. 10X Genomics Whole Genome Sequencing Analysis
R script (`titanCNA_v1.15.0_TenX.R`) for running TitanCNA analysis on sequencing data with haplotype phasing such as 10X Genomics whole genome data

**Input files**  
  1. GC-corrected, normalized read coverage using the HMMcopy suite  
    a. Get molecule coverage by counting unique barcodes in non-overlapping bins.  To do this, you will need the following installed:
      * [bxtools](https://github.com/walaj/bxtools#tile)
      * samtools
    To extract the molecule coverage at 10kb bins across chromosome 1 for both tumour and normal samples:
      ```
      samtools view -h -F 0x4 -q 60 tumour.bam 1 | bxtools tile - -b chr1.10kb.bed > tumour_bxTile/chr1.10kb.bxTile.bed
    samtools view -h -F 0x4 -q 60 normal.bam 1 | bxtools tile - -b chr1.10kb.bed > normal_bxTile/chr1.10kb.bxTile.bed
      ```
      The chromosome bed files for 10kb bins is provided in `TenX_scripts/data/10kb.bed.tar.gz`. Just `tar xvzf 10kb.bed.tar.gz` to extract the files.
    
    b. Normalize the molecule coverage
      ```
      # from the command line
      >Rscript TenX_scripts/getMoleculeCoverage.R --help
      Usage: TenX_scriptsgetMoleculeCoverage.R [options]
      
      
      Options:
              -d DATADIR, --datadir=DATADIR
                      Data library path containing gc/map wig files
      
              -t TUMORBXDIR, --tumorBXDir=TUMORBXDIR
                      Path to directory containing tumor bed files for each chromosome containing BX tags.
      
              -n NORMALBXDIR, --normalBXDir=NORMALBXDIR
                      Path to directory containing normal bed files for each chromosome containing BX tags.
      
              --minReadsPerBX=MINREADSPERBX
                      Minimum number of reads per barcode.
              
              (... plus other options)
      ```
      Here is an example 
      ```
      Rscript correctBXcounts_purity_tumNorm_v1.R --id test --datadir TenX_scripts/data/ \
        --tumorBXDir tumour_bxTile/ --normalBXDir normal_bxTile/ --minReadsPerBX 2 \
        --chrs "c(1:22, \"X\")" --outDir ../ \
        --centromere TenX_scripts/data/GRCh37.p13_centromere_UCSC-gapTable.txt \
        
      ```
  
  2. Tumour sample allelic read counts at heterozygous SNPs (identifed from the matched normal sample).
    a. Identify the phased heterozygous SNPs (excluding indels) from the LongRanger 2.1 analysis on the matched normal sample.  
      Use this R script from the command line to process the phased variant file (`*phased_variants.vcf.gz`) and extracts heterozygous sites after filtering.  The output VCF file suffix `*_phasedHets.vcf`
      ```
      # from the command line
      > Rscript TenX_scripts/getPhasedHETSitesFromLLRVCF.R --help
      Usage: Rscript getPhasedHETSitesFromLLRVCF.R [options]
      
      
      Options:
              -i INVCF, --inVCF=INVCF
                      LongRanger 2.1 phased variant result VCF file; typically has filename suffix "*phased_variants.vcf.gz". [Required]
      
              -q MINQUALITY, --minQuality=MINQUALITY
                      Heterozygous variants with QUAL greater than or equal to this value are considered. [Default: 100]
      
              -d MINDEPTH, --minDepth=MINDEPTH
                      Heterozygous variants with read depth greater than or equal to this value are considered. [Default: 10]
      
              -v MINVAF, --minVAF=MINVAF
                      Heterozygous variants with variant allele fraction or reference allele fraction greater than this value are considered. [Default: 0.25]
      
              -o OUTVCF, --outVCF=OUTVCF
                      Output VCF file with suffix "_phasedHets.vcf" [Required]
      
              -h, --help
                      Show this help message and exit
      ```
    
    b. Extract allelic read counts from the tumour sample
      Using this python script, extract the allele read counts at the heterozygous SNP sites identified in the previous step.  This script is meant to be run per chromosome so that users can parallelize this step.  Users will need to combine the chromosome files into a single tab-delimited file for input into the R script.
      ```
      # from the command line
      # run per chromosome
      # The `pysam` library will need to be installed in python.
      
      usage: python TenX_scripts/getTumourAlleleCountsAtHETSites.py $chr $phasedHetsVCF $bam $refFasta $baseQuality $mapQuality $vcfQuality > $tumCountsFile
      
      # $chr - Chromosome to analyze (e.g. 1)
      # $phasedHetsVCF - Output VCF file from getPhasedHETSitesFromLLRVCF.R
      # $bam - BAM file path of the tumour sample
      # $refFasta - File path to the reference FASTA file
      # $baseQuality - Minimum base quality to include in counts
      # $mapuality - Minimum mapping quality of reads to include in counts
      # $vcfQuality - Exclude HET sites with lower QUAL than this value
      # $tumCountsFile - Output tab-delimited file 
      ```
**Running the R script**  
1. Look at the usage of the R script  
  ```
  # from the command line
  > Rscript titanCNA_v1.15.0_TenX.R --help
  Usage: Rscript titanCNA_v1.15.0_TenX.R [options]


  Options:
        --id=ID
                Sample ID

        --hetFile=HETFILE
                File containing allelic read counts at HET sites. (Required)
        
        (... other arguments similar to titanCNA.R)
        
        --alleleModel=ALLELEMODEL
                Emission density to use for allelic input data (binomial or Gaussian). [Default: binomial]
                
        --alphaR=ALPHAR
                Hyperparaemter on the Gaussian variance for allelic fraction data; used if --alleleModel="Gaussian". [Defaule: 5000]
        
        --haplotypeBinSize=HAPLOTYPEBINSIZE
                Bin size for summarizing phased haplotypes. [Default: 100000]

        --phaseSummarizeFun=PHASESUMMARIZEFUN
                Strategy to summarize specified haplotype bins: mean, sum, SNP. [Default: sum]
  ```
  The specific arguments that are different compared to `titanCNA.R` are listed above.  `HETFILE` should be the output of `getTumourAlleleCountsAtHETSites.py`. 
  
2. Example usage of R script
  ```
  # normalized coverage file: test.cn.txt
  # allelic read count file from getTumourAlleleCountsAtHETSites.py: test.het.txt
  Rscript titanCNA.R --id test --hetFile test.het.txt --cnFile test.cn.txt \
    --numClusters 1 --numCores 1 --normal_0 0.5 --ploidy_0 2 \
    --chrs "c(1:22, \"X\")" --estimatePloidy TRUE --outDir ./ \
    --haplotypeBinSize 1e5 --phaseSummarizeFun sum --alleleModel Gaussian --alphaR 5000
  ```
  
3. Running TitanCNA for multiple restarts and model selection
  Use the same approach as for Step 3 of [Standard Whole Genome/Exome Sequencing Analysis](#wgs).


